### PR TITLE
[SPARK-24338][SQL] Fixed Hive CREATETABLE error in Sentry-secured cluster

### PR DESCRIPTION
**Pull request by yuchaoran2011**
## What changes were proposed in this pull request?

As detailed in the JIRA [ticket](https://issues.apache.org/jira/browse/SPARK-24338), Hive "CREATE TABLE" statement would fail when Sentry is used to manage authorization. This PR fixed this bug by porting the Cloudera fix.

## How was this patch tested?

Tested that the table can be successfully created with the fix.

### CODE REVIEW

**Comment by skonto** [comment]
@vanzin Could you pls review it? I know the fix exists on Cloudera's fork but right now since this is missing from the upstream project, it will affect all sentry users who dont use Cloudera's distro.

**Comment by vanzin** [comment]
Oh man, this thing. I didn't post this upstream because I think this is kinda hacky, and I also consider this primarily a Sentry bug. But if others are ok with it...

Just make sure you have all the necessary changes. My initial patch caused a couple of breakages.
https://github.com/cloudera/spark/commit/4c8289d
https://github.com/cloudera/spark/commit/629b1c3
https://github.com/cloudera/spark/commit/41fc8a1

Also you probably shouldn't reference CDH bugs in the code here...

**Comment by yuchaoran2011** [comment]
Thanks @vanzin. I've verified this commit has all the changes you mentioned. I've also removed references to CDH in the code comments.

**Comment by vanzin** [comment]
ok to test

**Comment by vanzin** [comment]
(Although I'm pretty sure it will fail style checks.)

Also, @cloud-fan since you wrote the original change.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 1]
```diff
@@ -230,11 +232,29 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
...
-    val needDefaultTableLocation = tableDefinition.tableType == MANAGED &&
-      tableDefinition.storage.locationUri.isEmpty
-
-    val tableLocation = if (needDefaultTableLocation) {
-      Some(CatalogUtils.stringToURI(defaultTablePath(tableDefinition.identifier)))
+    //
+    // When using a remote metastore, and if a managed table is being created with its
+    // location explicitly set to the location where it would be created anyway, then do
+    // not set its location explicitly. This avoids an issue with Sentry in secure clusters.
+    // Otherwise, the above comment applies.
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: We have comments saying
```
// We can't leave `locationUri` empty and count on Hive metastore to set a default table
// location, because Hive metastore uses hive.metastore.warehouse.dir to generate default
// table location for tables in default database, while we expect to use the location of
// default database.
```

In general I think it's OK to always set the location for managed tables, and I feel a little weird if some systems forbid it. Hive supports setting a location for managed tables, how shall we handle that?

Comment (vanzin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: I'm not sure I really follow that comment. IIRC you can't change the location of the default database in Hive, so I'm not sure how you'd hit a situation where "default db uri != warehouse dir", at least when using a remote metastore.

Comment (vanzin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: > In general I think it's OK to always set the location for managed tables

BTW I don't disagree with that and as I said I believe this is ultimately a bug in Sentry, not in Spark. The question is whether we want this workaround in Spark while Sentry is not fixed.

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: > IIRC you can't change the location of the default database in Hive

good to know that, maybe we should apply the same rule to Spark.

Comment (skonto) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: >  The question is whether we want this workaround in Spark while Sentry is not fixed.
We could open a bug for Sentry and since Spark uses hive which often comes with Sentry in enterprise envs then it would be good to point to that bug in the code , so when things get fixed we could change bug the code here.

**Comment by skonto** [comment]
@vanzin should we merge this or add a comment in the docs?

**Comment by vanzin** [comment]
I'm kinda neutral on what should happen, except maybe for filing a Sentry bug. I'll let @cloud-fan decide whether this makes sense in Spark.

For the Sentry bug, there's an easy way to reproduce this, just run the following (either in Spark or using the old Hive CLI - not beeline nor jdbc):

```
USE some_db;
CREATE TABLE foo(idx int) LOCATION 'hdfs://blah.example.com/user/hive/warehouse/some_db.db/foo';
```

Even if the user has permissions to create table on that database, Sentry will not allow it. Works fine if you remove the explicit location.

**Comment by yuchaoran2011** [comment]
@vanzin But in Spark SQL, even when LOCATION is not specified, Sentry would still not allow the table to be created.

**Comment by vanzin** [comment]
If that's the case then your user does not have permission to create the table. The issue that my patch works around is a bug in Sentry where creating a table *with the location specified to the same location as it would have been automatically assigned* does not work. My patch removes the explicit location in that case, and things work.

But if you don't have permissions, you'll get an error regardless of my patch.

**Comment by yuchaoran2011** [comment]
I can double check. But I remember clearly that the user did have all the necessary privileges. Without this patch, a permission error was thrown. With this patch it worked fine. The CREATE TABLE statement didn't involve a LOCATION.

**Comment by vanzin** [comment]
Yes, that's exactly what I said. I'm not sure what part is confusing you.

**Comment by vanzin** [comment]
You said: "even when LOCATION is not specified, Sentry would still not allow the table to be created."

I said that if that's the case, then your user doesn't have the needed permission. Because all my patch does is remove the location.

I'm still confused about what it is that you're confused about.

**Comment by yuchaoran2011** [comment]
I'm not confused about anything.

**Comment by vanzin** [comment]
I think you were confused by my example. It was not meant to be a workaround for Spark users (since currently Spark will just add the location if it's not there), it was meant to be used if creating the Sentry bug report ("this command should work but fails, and the corresponding command without the explicit location works").

**Comment by cloud-fan** [comment]
I don't think working around a bug in Sentry is a good reason to add a hack in Spark. `HiveExternalCatalog` is already full of hacks...

If you have a proposal to clean that up and can also work around the Sentry bug, that will be great.

**Comment by skonto** [comment]
@vanzin @cloud-fan. Going back to the facts. The fix/work-around in Cloudera's distro obviously exists because the distro integrates with Sentry anyway. Removing the location did fix the Spark Job at the customer side. To re-phrase the last request from @cloud-fan, if we remove the explicit location what are the consequences, when do you need the explicit location for managed tables? Can this be a default behavior in general?

**Comment by blue666man** [comment]
To weigh in a bit, we're one of the larger enterprises that hit this bug, which we consider a regression going from Spark 1.6.3 (on Mesos, attempting to perform DDL to a Hive instance on CDH) to >= 2.1.x.  I think many more large enterprises will hit this with Spark on K8S to any Sentry-secured Hive databases.

IMO, [this commit](https://github.com/apache/spark/commit/ce13c2672318242748f7520ed4ce6bcfad4fb428), while well intended in removing some of the hacks for 2.1.0, was a reversal of Spark's sort-of-implicit-commitment to have Sentry/Hive DDL work.  I agree this should be fixed upstream in Sentry too, but that would be a long wait for us:  creating tables in Mesos Spark 2.x is blocking 100's of devs from upgrading to Spark 2.

Basically, the hacks / workarounds had existed in the past and were working.  Please add a workaround back until Sentry can fix the location issue on their side.

**Comment by vanzin** [comment]
I don't disagree with what you guys are saying, but I also don't have the context of why the original change was made.

**Comment by skonto** [comment]
@vanzin any decision on this?

**Comment by vanzin** [comment]
I didn't make the original change so I'm refraining from making a decision about whether this should go in. Perhaps another fix is desired, perhaps the current behavior is desired, but either way this isn't an area where I'm comfortable making these kinds of decisions.

(CDH is another thing altogether.)
